subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskRequiredServices.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks;
+
+import org.gradle.api.internal.TaskInternal;
+import org.gradle.api.internal.provider.DefaultProperty;
+import org.gradle.api.internal.tasks.properties.PropertyVisitor;
+import org.gradle.api.internal.tasks.properties.PropertyWalker;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.services.BuildService;
+import org.gradle.api.services.BuildServiceRegistry;
+import org.gradle.api.services.internal.BuildServiceProvider;
+import org.gradle.internal.Cast;
+import org.gradle.internal.reflect.validation.TypeValidationContext;
+
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+public class DefaultTaskRequiredServices implements TaskRequiredServices {
+    private final TaskInternal task;
+    private final TaskMutator taskMutator;
+    private final PropertyWalker propertyWalker;
+    private final BuildServiceRegistry buildServiceRegistry;
+    private Set<Provider<? extends BuildService<?>>> servicesRegistered;
+    private Set<Provider<? extends BuildService<?>>> servicesRequired;
+
+    public DefaultTaskRequiredServices(TaskInternal task, TaskMutator taskMutator, PropertyWalker propertyWalker, BuildServiceRegistry buildServiceRegistry) {
+        this.task = task;
+        this.taskMutator = taskMutator;
+        this.propertyWalker = propertyWalker;
+        this.buildServiceRegistry = buildServiceRegistry;
+    }
+
+    @Override
+    public Set<Provider<? extends BuildService<?>>> getRequiredServices() {
+        if (servicesRequired == null) {
+            servicesRequired = collectServicesRequired();
+        }
+        return servicesRequired;
+    }
+
+    @Override
+    public void visitRegisteredProperties(PropertyVisitor visitor) {
+        Set<Provider<? extends BuildService<?>>> tmpServicesRegistered = servicesRegistered;
```

Comment:
🎨 A name pun here would be neater (`servicesRegistered = this.servicesRegistered`).